### PR TITLE
Refactoring Add script to count unique learners in Sensei

### LINKED ISSUE #2111: Script to count unique learners
## Goal
We're looking to track the increase in learners we expect will occur when we launch Learning Pathways next year. To do this, we need to see the number of unique course subscribers (called "students" in Sensei) on Learn.

## Problem
Theoretically, this should be shown on https://learn.wordpress.org/wp-admin/admin.php?page=sensei_reports . However, due to the way bbPress is integrated with make.wordpress.org, we are only seeing a small subset there. This can be confirmed by going to https://learn.wordpress.org/wp-admin/admin.php?page=sensei_reports&course_id=8657&view=user, which shows we have at least 8,900 learners subscribed to that course alone.

## Solution
@m1r0 has written us a script that will count the number of unique learners on Learn and display this on [the Students report screen](https://learn.wordpress.org/wp-admin/admin.php?page=sensei_reports). It also includes a date filter which will only count students who started a course in that time period.

```
add_action( 'sensei_reports_overview_before_top_filters', 'wporg_learn_add_student_count_to_reports' );
function wporg_learn_add_student_count_to_reports( $type ) {
    if ( 'users' !== $type ) {
        return; // Only show the count on the students report screen.
    }
 
    $from_date = DateTime::createFromFormat( 'Y-m-d', $_GET['from_date'] ?? '', new DateTimeZone( 'UTC' ) );
    $to_date   = DateTime::createFromFormat( 'Y-m-d', $_GET['to_date'] ?? '', new DateTimeZone( 'UTC' ) );
 
    global $wpdb;
    $query = "SELECT COUNT(DISTINCT user_id)
FROM $wpdb->comments
WHERE comment_type = 'sensei_course_status'";
 
    if ( $from_date ) {
        $from_date->setTime( 0, 0, 0 );
 
        $query .= " AND comment_date_gmt >= '" . $from_date->format( 'Y-m-d H:i:s' ) . "'";
    }
 
    if ( $to_date ) {
        $to_date->setTime( 23, 59, 59 );
 
        $query .= " AND comment_date_gmt <= '" . $to_date->format( 'Y-m-d H:i:s' ) . "'";
    }
 
    $student_count = $wpdb->get_var( $query );
 
    ?>
    <div class="actions bulkactions">
        <label><?php esc_html_e( 'Total number of students', 'wporg-learn' ); ?></label>
        <input
            class="sensei-date-picker"
            name="from_date"
            type="text"
            autocomplete="off"
            placeholder="<?php echo esc_attr( __( 'From Date', 'sensei-lms' ) ); ?>"
            value="<?php echo esc_attr( $from_date ? $from_date->format( 'Y-m-d' ) : '' ); ?>"
        />
        <input
            class="sensei-date-picker"
            name="to_date"
            type="text"
            autocomplete="off"
            placeholder="<?php echo esc_attr( __( 'To Date', 'sensei-lms' ) ); ?>"
            value="<?php echo esc_attr( $to_date ? $to_date->format( 'Y-m-d' ) : '' ); ?>"
        />
        <label>: <?php echo (int) $student_count; ?></label>
    </div>
    <br>
    <?php
}
```

image-13

## Next Steps
What is the best way to get this script deployed on Learn? (CC: @adamwoodnz )
//dev

### PR DESCRIPTION
**Pull request by jonathanbossenger**
Update to #2172
Fixes #2111 
Add sanitization for query strings
Use wpdb->prepare for SQL query

### CODE REVIEW

**Comment by jonathanbossenger** [comment]
@bsanevans In attempting to refactor #2172 to include sanitizing the query variables and to use $wpdb->prepare for the query, I had to make quite a few changes to the code.

Could I ask you to get  @m1r0 to review this refactored version to confirm it would still work functionally the same as the original code, then I'll get someone else from the meta team to review it.

**Comment by kaitohm** [comment]
I've reached out to @m1r0 .

File: wp-content/plugins/wporg-learn/inc/sensei.php [hunk 1]
```diff
@@ -163,3 +163,81 @@ function wporg_fix_learning_mode_header_space() {
...
+ * Get the number of unique learners between two dates.
+ *
+ * @param \DateTime $from_date
+ * @param \DateTime $to_date
+ *
+ * @return int
+ */
+function wporg_learn_get_student_count( $from_date, $to_date ) {
+
+	if ( ! $from_date && ! $to_date ) {
```
Comment (m1r0) [inline wp-content/plugins/wporg-learn/inc/sensei.php]: I'm getting an error when filling in only one of the dates.

```
Call to a member function format() on bool
```

Do you intend to make it that both dates are required? If that's the case, maybe the logic should be:

```suggestion
	if ( ! $from_date || ! $to_date ) {
```

WDYT?

**Review by m1r0 — COMMENTED** [review]
Hello, hello! 👋 🙂 

> still work functionally the same as the original code

The only functional difference I'm noticing is that now both dates are needed. If that's intended, I've left some feedback.

Everything else looks solid. 🪨

**Comment by kaitohm** [comment]
It would be great if we didn't need to set both dates. For example, if we could get all data from a specific date until today without needing to set today's date, that would be great.

@jonathanbossenger Is that a change you can implement?

File: wp-content/plugins/wporg-learn/inc/sensei.php (hunk 1 above)
Comment (jonathanbossenger) [reply wp-content/plugins/wporg-learn/inc/sensei.php]: @m1r0 thanks for the feedback. My focus was on the refactoring to get the php linter to pass, so I wasn't 100% sure myself, happy to make the change.

**Comment by jonathanbossenger** [comment]
Thanks @m1r0 and @bsanevans I've updated the PR.

@pkevan This PR is an update to #2172, adding sanitization to the date variables, and changing the SQL concatenation to use a prepared statement.

Could we ask you to give this a review?

Thanks

**Review by kaitohm — APPROVED** [review]

**Review by pkevan — APPROVED** [review]
👍 Looks good

**Pull request merged**